core/shared/src/main/scala/scalaz/zio/Repeat.scala

```diff
@@ -0,0 +1,483 @@
+          z     <- fz(self.value(state))
+        } yield (state, z)
+
+      val start = self.start
+
+      val value = (state: State) => state._2
+
+      val update = (a: A, s0: State) =>
+        for {
+          step <- self.update(a, s0._1)
+          z    <- f(s0._2, self.value(step.value))
+        } yield step.map(s => (s, z))
+    }
+
+  /**
+   * Returns a new schedule that reduces over this one.
+   */
+  final def reduce[B1 >: B](f: (B1, B1) => B1): Repeat[A, B1] =
+    fold[B1](identity)(f)
+
+  /**
+   * Returns the composition of this schedule and the specified schedule,
+   * by piping the output of this one into the input of the other, and summing
+   * delays produced by both.
+   */
+  final def >>>[C](that: Repeat[B, C]): Repeat[A, C] =
+    new Repeat[A, C] {
+      type State = (self.State, that.State)
+      val initial = (a: A) => self.initial(a).flatMap(s1 => that.initial(self.value(s1)).map(s2 => (s1, s2)))
+      val start   = self.start.parWith(that.start)(_ + _)
+      val value   = (state: State) => that.value(state._2)
+      val update = (a: A, s: State) =>
+        self.update(a, s._1).flatMap { step1 =>
+          if (step1.cont) that.update(self.value(step1.value), s._2).map { step2 =>
+            step1.combineWith(step2)(_ && _, _ + _)
+          } else IO.now(Step.done((step1.value, s._2), step1.delay))
+      }
+    }
+
+  /**
+   * A backwards version of `>>>`.
+   */
+  final def <<<[C](that: Repeat[C, A]): Repeat[C, B] = that >>> self
+
+  /**
+   * An alias for `<<<`
+   */
+  final def compose[C](that: Repeat[C, A]): Repeat[C, B] = self <<< that
+}
+
+object Repeat {
+  sealed case class Step[+A](cont: Boolean, value: A, delay: Duration) { self =>
+    final def map[B](f: A => B): Step[B] = copy(value = f(value))
+
+    final def delayed(f: Duration => Duration): Step[A] = copy(delay = f(delay))
+
+    final def combineWith[B](that: Step[B])(g: (Boolean, Boolean) => Boolean,
+                                            f: (Duration, Duration) => Duration): Step[(A, B)] =
+      Step(g(self.cont, that.cont), (self.value, that.value), f(self.delay, that.delay))
+  }
+  object Step {
+    def cont[A](a: A, d: Duration): Step[A] = Step(true, a, d)
+    def done[A](a: A, d: Duration): Step[A] = Step(false, a, d)
+  }
+
+  final def apply[A, B](initial0: A => IO[Nothing, B],
+                        start0: IO[Nothing, Duration],
+                        update0: (A, B) => IO[Nothing, Repeat.Step[B]]): Repeat[A, B] = new Repeat[A, B] {
+    type State = B
+    val initial = initial0
+    val start   = start0
+    val value   = Predef.identity
+    val update  = update0
+  }
+
+  /**
+   * A schedule that repeats forever, returning each input as the output.
+   */
+  final def identity[A]: Repeat[A, A] =
+    Repeat[A, A](IO.now(_), IO.now(Duration.Zero), (a, _) => IO.now(Step.cont(a, Duration.Zero)))
+
+  /**
+   * A schedule that never executes.
+   */
+  final def never: Repeat[Any, Unit] =
```

Comment:
I wonder if it is possible to write a `Repeat[Any, Nothing]`, I guess not without reworking `Step`
```scala
Repeat[Any, Nothing](_ => IO.never, IO.never.const(Duration.Inf), ???)
```